### ISSUE TITLE
`docker-machine ssh` failing without helpful error message

### ISSUE THREAD

**Issue by metasim** [docker/machine#3349]
I can directly `ssh` to my server with password free key pair, both as root and myself. However, when I try to use `docker-machine ssh myserver` I get an ssh usage error:

```
$ docker-machine -D ssh devops03
Docker Machine Version:  0.7.0, build a650a40
Found binary path at /usr/local/bin/docker-machine
Launching plugin server for driver none
Plugin server listening at address 127.0.0.1:57992
() Calling .GetVersion
Using API Version  1
() Calling .SetConfigRaw
() Calling .GetMachineName
(devops03) Calling .GetState
(devops03) Calling .GetSSHHostname
(devops03) Calling .GetSSHPort
(devops03) Calling .GetSSHKeyPath
(devops03) Calling .GetSSHUsername
Using SSH client type: external
&{[-F /dev/null -o BatchMode=yes -o PasswordAuthentication=no -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=quiet -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none @ -p 0] /usr/bin/ssh <nil>}
&{/usr/bin/ssh [/usr/bin/ssh -F /dev/null -o BatchMode=yes -o PasswordAuthentication=no -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=quiet -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none @ -p 0] []  <nil> <nil> <nil> [] <nil> <nil> <nil> <nil> false [] [] [] [] <nil>}
usage: ssh [-1246AaCfGgKkMNnqsTtVvXxYy] [-b bind_address] [-c cipher_spec]
           [-D [bind_address:]port] [-E log_file] [-e escape_char]
           [-F configfile] [-I pkcs11] [-i identity_file]
           [-L [bind_address:]port:host:hostport] [-l login_name] [-m mac_spec]
           [-O ctl_cmd] [-o option] [-p port]
           [-Q cipher | cipher-auth | mac | kex | key]
           [-R [bind_address:]port:host:hostport] [-S ctl_path] [-W host:port]
           [-w local_tun[:remote_tun]] [user@]hostname [command]
exit status 255
```

When  I attempt to use the "native" ssh client I get this:

```
$ docker-machine -D --native-ssh ssh devops03
Docker Machine Version:  0.7.0, build a650a40
Found binary path at /usr/local/bin/docker-machine
Launching plugin server for driver none
Plugin server listening at address 127.0.0.1:58256
() Calling .GetVersion
Using API Version  1
() Calling .SetConfigRaw
() Calling .GetMachineName
(devops03) Calling .GetState
(devops03) Calling .GetSSHHostname
(devops03) Calling .GetSSHPort
(devops03) Calling .GetSSHKeyPath
(devops03) Calling .GetSSHUsername
Using SSH client type: native
&{{{<nil> 0 [] [] []}  [] <nil>  []}  0 <nil>}
dial tcp :0: connect: can't assign requested address
```

Is thee some why of getting more information on where the problem is?

**Comment by nathanleclaire** [comment]
However you set this machine's info, it seems to not be persisted correctly.  Notice that the invoked SSH command doesn't have an address or user, just a bare "@":

``
&{/usr/bin/ssh [/usr/bin/ssh -F /dev/null -o BatchMode=yes -o PasswordAuthentication=no -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=quiet -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none @ -p 0] []  <nil> <nil> <nil> [] <nil> <nil> <nil> <nil> false [] [] [] [] <nil>} 
```

**Comment by metasim** [comment]
Yep, that's the problem:

```json
...
    "ConfigVersion": 3,
    "Driver": {
        "IPAddress": "192.168.4.18:2376",
        "MachineName": "devops03",
        "SSHUser": "",
        "SSHPort": 0,
        "SSHKeyPath": "",
        "StorePath": "/Users/sfitch/.docker/machine",
        "SwarmMaster": false,
        "SwarmHost": "",
        "SwarmDiscovery": "",
        "URL": "tcp://192.168.4.18:2376"
    },
    "DriverName": "none",
...
```

I went and edited the `config.json` file by hand, but it didn't stick. Would that be expected?

**Comment by nathanleclaire** [comment]
OK, so it seems that the fields are getting over-written somehow !  Unfortunate.

Do you have any minimal dupe, or at least more detail on the steps how you ended up in this situation?

**Comment by metasim** [comment]
I created the entries using the `none` driver, based on this script:

https://github.com/docker/machine/issues/3344#issuecomment-212536797

**Comment by metasim** [comment]
PS: Note I updated the original followup with more details (there were errors in my first characterization).

**Comment by nathanleclaire** [comment]
Oh, if they're created with `none`, I wouldn't really expect anything related to SSH to work.  Why not use `generic`?

**Comment by nathanleclaire** [comment]
Ah, I see in your original issue, you reference the "regeneration of certs" dilemma.  Instead of using `none` for this, maybe we could make the driver or provisioning process a bit more robust to pull down existing certificates if they are already present and valid?

**Comment by metasim** [comment]
I certainly have no affinity for the `none` driver, but I can't get the `generic` driver to work without it a) messing with he existing certs, b) running yum/apt update and messing things up there, or c) messing with the hostname. Stuck in this place where `none` doesn't do enough, and `generic` does too much. I think issue #1221 captures this.

**Comment by nathanleclaire** [comment]
> Stuck in this place where none doesn't do enough, and generic does too much.

Yeah, I hear you.  It seems that the most likely solution would be to copy the certs down from an instance during the provisioning process if they already exist instead of blindly regenerating new ones.  Would that fix your issue?

(Side note: It occurs to me that an additional difficulty here might be the over-writing of configuration specified with `--engine-*` type options. Tricky :( )

**Comment by metasim** [comment]
> Yeah, I hear you. It seems that the most likely solution would be to copy the certs down from an instance during the provisioning process if they already exist instead of blindly regenerating new ones. Would that fix your issue?

Ideally, `docker-machine create --driver generic` would check to see if the docker daemon is already running, and just configure the local `docker-machine` config with the settings and certificates associated with the preexisting instance (assuming `root` ssh access is available/configured). If the server version is behind, instead of automatically calling `yum` or `apt` update, it would issue as warning, but not automatically run update (or change anything else).

IOW, if `create --driver generic` is run against a machine with no docker installed, then do all the automatic provisioning. Otherwise, take a more conservative approach.

**Comment by nathanleclaire** [comment]
Yup, that all sounds good to me.

**Comment by rhaarm** [comment]
What this ever accomplished, does `--driver generic` have a way to skip on install, restart of the existing docker daemon and just only pull down the existing certificates? I'm trying to share the docker swarm manager with a co-worker and the only way to get them access to the machine is through `--driver none`. Which with the `--driver none` I have to copy all machine certs on my machine and send them over to them. It would be nice there was a way if he already has an existing login credentials to the Host machine via SSH, there would be a way to connect and setup everything.
